sql-tuning-best-practice.md

```diff
@@ -0,0 +1,714 @@
+
+```SQL
+SELECT id, name FROM emp WHERE id = 901; 
+```
+
+### Logical Transformation
+
+The purpose of logical Transformation is to optimize the execution of statements based on the characteristics of SELECT list, WHERE predicates, and other predicates in SQL queries. It generates a logical execution plan to annotate and rewrite the query. This logical plan is then passed to the Cost-Based Optimization. The optimization rules include column pruning，partition pruning，eliminate Max/Min, eliminate outer join, join reorder, predicates push-down，subquery rewrite，derive TopN from window functions，and de-correlation of correlated Subquery. Since this step is fully automated by the query optimizer, it usually does not require manual adjustments.
+
+More Detail for Logical Transformation: https://docs.pingcap.com/tidb/stable/sql-logical-optimization.
+
+### Cost-Based Optimization
+
+TiDB uses statistics as input to the optimizer to estimate the number of rows processed in each plan step for a SQL statement, and associates a cost with each plan step. The Cost-Based Optimization estimates the cost of each available plan choice, including index accesses and the sequence of table joins, and produces a cost for each available plan. The optimizer then picks the execution plan with the lowest overall cost.
+
+The below figure illustrates the various data access paths and row set operations that cost-based optimization can consider to develop the optimal execution plan. Furthermore, during the logical transformation stage, the query has already been rewritten for predicate push-down. In the cost-based optimization stage, TiKV expression push-down is further implemented when possible at TiKV layer. 
+
+Furthermore, confirming the algorithm for certain SQL operations, such as aggregation, join, and sorting, is essential. For instance, the aggregation operator may utilize either HASH_AGG or STREAM_AGG, while the join operator can select from HASH JOIN, MERGE JOIN, or INDEX JOIN. Likewise, various options are available for the sorting operator.
+
+![cost-based-optimization](/media/sql-tuning/cost-based-optimization.png)
+
+## Understanding Execution Plans
+
+The execution plan represents the steps TiDB will follow to execute a SQL query. In this section, we will learn how to display and read the execution plan.
+
+### Generating and Displaying Execution Plans
+
+Beside access the execution plan information through TiDB Dashboard, TiDB provides a `EXPLAIN` statement to display the execution plan for a SQL query. Here's an example of using `EXPLAIN`:
+- id: Operator name and the step unique identifier
+- estRows: Estimated number of rows from the particular step
+- task: The executor of the step
+- access object: The object where the row sources are located
+- operator info: Extended information about the operator regarding the step
+
+```SQL
+tidb> EXPLAIN SELECT count(*) FROM trips WHERE start_date BETWEEN '2017-07-01 00:00:00' AND '2017-07-01 23:59:59';
```

Comment:
```suggestion
EXPLAIN SELECT COUNT(*) FROM trips WHERE start_date BETWEEN '2017-07-01 00:00:00' AND '2017-07-01 23:59:59';
```

For this and other output:
- Set the language to `sql` instead of `SQL`
- Only the query is valid SQL, not the output
- The prompt should not be in there and if it is in there it should probably be `mysql> ` as that's what users would see without additional configuration.